Add handling of policy metadata

jjbuchan opened the pull request:
# What

1. Adds the utilization of policy metadata for monitors that are created/updated with no custom value being set on any of the metadata fields.
1. Adds the handling of policy metadata change events which are sent by Policy Mgmt.

# How

## Monitor Creates
Whenever monitors are created any `null` field that has the `@MetadataPolicy` annotation will use the corresponding Policy Metadata value if it exists.  These values will be retrieved via the `PolicyApi`.

We have to handle this in two different places.
1. The plugin contents are stored on the Monitor as a String, so we must set any metadata fields for the plugins within the `MonitorConversionService`.  These contents will be passed to the `MonitorManagement` service via the `MonitorCU` object generated.
1. The top level Monitor fields using policy metadata will be set within MonitorManagement.

Since both of these use similar logic, a `MetadataUtils` class was created to assist with that.

## Monitor Updates
For update operations, the values using metadata will no longer be set as `null` so we will instead look to see which metadata fields were previously set on the Monitor (these are stored in two different lists - one for plugin metadata and one for monitor metadata), and compare the values of what was received against the currently policies.  If those values match then we will assume the value provided is still supposed to be using metadata.  if the values differ, that field will be removed from the list of metadata fields utilized on that monitor/plugin.

---

## Metadata Change Events

Whenever Policy Metadata changes, an event will be sent from PolicyMgmt and consumed by MonitorMgmt, which will lead to the Monitor object being updated with the new values.

The relevant monitors can be found by querying the Monitors table for monitors that are utilizing the metadata field (based on the aforementioned lists of utilized metadata fields that get stored on each monitor object).

For account monitors, the bound monitors will then be unbound then re-bound.  For policy monitors, an event will be sent to trigger the same process that happens when a Policy Monitor is updated directly via the API.  i.e. an event goes to PolicyMgmt who determines the relevant tenants and sends an event for each one, which will be consumed again by MonitorMgmt and lead to an unbind/re-bind.

---

**Note:** Since pushing this PR it has been decided that we'll actually change the logic to do the inverse.  Instead of specifying `@MetadataPolicy` on fields that can use metadata, we will instead specifying (something like) `@NotMetadataPolicy` to those that cannot use it.  This will allow us to easily change default monitor configuration at any time in the future.  

e.g. instead of a Cpu class containing
```
@JsonProperty(defaultValue = "true")
  boolean totalcpu = true;
```
that value will instead be stored in a Metadata Policy in case we want to alter it in future.

Assuming we all agree on this, that change will come in a future PR.

## How to test

Run tests.
Manually.

# Why

When monitors are created without custom values, we should set those values to use Rackspace defaults.  Rather than just storing the value we also store the fact the monitor is using the default so that if the default value changes at any time in the future, those monitors using it will also be updated.

Whenever policy metadata (the default configuration settings for monitors/tasks) is changed we need to be able to find which monitors are currently using that metadata and update them as well as all the associated bound monitors.

# TODO

1. Better handling of Monitoring Zones.  This PR allows us to create a policy for zones but it does not allow us to create different policies based on the region of the resource being monitored, which is what we want.
1. Better handling of policy metadata across account types for policy monitors.  This currently works as we'd want for account monitors using policy metadata, but for policy monitors it assumes they will use metadata stored at the global scope.
   * This is because we are storing the monitor object with the metadata values stored within it, which means one policy monitor will be configured the same for all tenants regardless of their account type.
   * There are a couple options here, but none of the available ones are likely ideal.
1. MetadataUtils could be moved out of here and into a common place for the EventEngineMgmt to also utilize.
   * Rather than using `MonitorMetadataPolicyDTO` it could use the parent `MetadataPolicyDTO`
   * I held off on doing that here because it added more work and I wanted to get this out first.
      * We'd have to convert a map of `MonitorMetadataPolicyDTO` into a map of `MetadataPolicyDTO`
      * Would also need to decide where this Utils class belongs.
          * Being part of PolicyMgmt probably makes sense?
          * Could maybe allow it to be imported in a similar way to what we do for API Clients?
1. More Insomnia workspace additions
   * PR incoming soon, I hope.
1. Should we set defaults for some `@NotNull` fields so that things don't fail if PolicyMgmt is down?  e.g. `interval` is now `@NotNull`.  That feels like a field we could have a general default for where it's acceptable to fallback to that even though it means it wont be using policy values.
   * It's probably worth waiting to see how things look before we bother setting any of these.
1. See the **Note** in the **How** section.  It may also clash with the above point.
1. Update the salus-docs with all this info

itzg commented on src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:
```diff
@@ -1655,4 +1809,48 @@ else if (m.getLabelSelectorMethod().equals(LabelSelectorMethod.OR)) {
+   * @param tenantId The tenant id the monitor is created under.
+   * @param monitor The monitor object to update.
+   */
+  void setMetadataFields(String tenantId, Monitor monitor) {
```
This seems quite similar to the new `MonitorConversionService.setMetadataFields`...both of which are complex implementations. It makes me wonder if `MetadataUtils` should be "promoted" to be a full blow Spring component/service and can encapsulate and co-locate these two similar, but necessarily different methods.

jjbuchan replied:
I looked at these a few times to see about consolidating them, but they were different in awkward places where it wouldn't be as easy to create a generic method.

Moving them over and keeping them two unique methods might be for the best.

GeorgeJahad replied:
If they can't be merged, then I would also like to see them in the same file, perhaps with a comment saying that any changes in one likely need to occur in the other.

jjbuchan replied:
I've moved them locally.  They'll be in the same file but unfortunately I think they have to stay separate.

jjbuchan commented on src/main/java/com/rackspace/salus/monitor_management/services/MonitorConversionService.java:
```diff
@@ -206,4 +219,44 @@ private void assertPluginAgentType(Monitor monitor, Object plugin) {
+
+    if (policyMetadata == null) {
+      // this api request is avoided if there are no metadata fields to set
+      policyMetadata = policyApi.getEffectiveMonitorMetadataMap(tenantId, className, monitor.getMonitorType());
```
It gets called at most once.  Either on line 239 or here.  

It will get called here if both the else is hit on 242 and 242 results in a non-empty response.

The simplification of this would be to always make this request on line 237, but avoiding the call where possible seemed beneficial even if it makes things a little less readable.

jjbuchan commented on src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:
```diff
@@ -128,7 +139,9 @@ public MonitorManagement(
       PolicyApi policyApi, ResourceApi resourceApi,
       ZoneManagement zoneManagement, ZonesProperties zonesProperties,
-      JdbcTemplate jdbcTemplate) throws IOException {
+      MonitorConversionService monitorConversionService,
```
I only added 2 :)

jjbuchan replied:
but in all seriousness, since the 2nd is object mapper I'm not sure I agree... but I see you have some other comments further below around this and may agree that moving some logic over to conversion service is better.

/me reads on

itzg replied:
Oh yeah, the insertion before `jdbcTemplate` threw me off :)

jjbuchan commented on src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:
```diff
@@ -1005,6 +1031,140 @@ void handleMonitorPolicyEvent(MonitorPolicyEvent event) {
+
+    Optional<MonitorMetadataPolicyDTO> policyOptional = effectivePolicies.stream()
+        .filter(p -> p.getId().equals(event.getPolicyId()))
+        .findAny();
```
> Is `effectivePolicies` expected to be only size one or zero?

That's true.

This (and the line below) are validating that the policy in the event is an active policy in use for this tenant.  There would only be 1 or 0 policies with that id in the effective policies list.

> I'm not sure why a `findAny` is used here.

I'm not sure what would be the alternative?  Collect them to a list? `findFirst`?

itzg replied:
Yeah, perhaps a combo of `findFirst` and a little more to the inline comment further up that clarifies a list is returned from `getEffectiveMonitorMetadataPolicies` but only one or zero should be in that list.

jjbuchan replied:
Realized my comment was slightly innaccurate.

`effectivePolicies` can be many policies, but once the filter is applied, only 1 or 0 should be found.

I used `findAny` because it is (at least can be) more efficient than `findFirst`.

itzg replied:
Oh...I had glossed over the fact the filter would narrow to one or zero. This kind of confusion is one reason I try to avoid the combined DB/REST queries followed by a filter-by-stream.

I'm fine with `findAny` now that I understand the full scenario.

GeorgeJahad replied:
I don't mind findAny, (although findFirst seems clearer to me,) but if the filtered list is only going to be 1 entry long, i wouldn't be too worried about efficiency.

GeorgeJahad replied:
also, i think i often create queries followed by filter-by-stream.  i actually like how clear they seem.  i guess i must think differently.

jjbuchan replied:
> but if the filtered list is only going to be 1 entry long

Good point.  I'll change it.

jjbuchan commented on src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:
```diff
@@ -1005,6 +1031,140 @@ void handleMonitorPolicyEvent(MonitorPolicyEvent event) {
+        // When the policy relates to plugin details the monitor must be converted
+        // so the Remote/LocalPlugin fields can be accessed.
+        DetailedMonitorInput input = new DetailedMonitorInput(
+            monitorConversionService.convertToOutput(monitor));
```
Yeah.  I don't really like this and it creates some potential problems.  I'm not sure what the solution would be though.  I might ping you for a chat once I'm through the other comments.

jjbuchan commented on src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:
```diff
@@ -1005,6 +1031,140 @@ void handleMonitorPolicyEvent(MonitorPolicyEvent event) {
+        }
+
+        try {
+          monitor.setContent(objectMapper.writeValueAsString(plugin));
```
I think I agree with this.

jjbuchan commented on src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:
```diff
@@ -1005,6 +1031,140 @@ void handleMonitorPolicyEvent(MonitorPolicyEvent event) {
+      // This will result in PolicyMgmt sending an event that will trigger processPolicyMonitorUpdate
+      // being ran for each tenant using the policy monitor.
+      for (Monitor m : relevantMonitors) {
+        monitorEventProducer.sendPolicyMonitorUpdateEvent(
```
I haven't created new diagrams for the metadata fields.  I'll hold off on that until we're able to chat a little.

...but it's a similar methodology as policy monitor update operations - https://github.com/Rackspace-Segment-Support/salus-docs-internal/pull/6/files#diff-90680e709513d53c13dee066ace77821

jjbuchan commented on src/main/java/com/rackspace/salus/monitor_management/utils/MetadataUtils.java:
```diff
@@ -0,0 +1,157 @@
+   * @return The list of metadata fields found.
+   */
+  public static List<String> getMetadataFieldsForCreate(Object object) {
+    List<String> unsetMetadataFields = new ArrayList<>();
```
I can see that now.  adjective vs. verb.  I'll think of a new name.

GeorgeJahad replied:
fwiw, i read it as an adjective, so it was clear to me.

jjbuchan replied:
I've changed it to `metadataFields` like the method below.

GeorgeJahad commented on src/main/java/com/rackspace/salus/monitor_management/services/MonitorConversionService.java:
```diff
@@ -110,7 +124,7 @@ public DetailedMonitorOutput convertToOutput(Monitor monitor) {
   }
 
-  public MonitorCU convertFromInput(DetailedMonitorInput input) {
+  public MonitorCU convertFromInput(String tenantId, UUID monitorId, DetailedMonitorInput input) {
```
do you think it would be clearer to call it "policyMonitorId"?

GeorgeJahad replied:
nevermind, i misunderstood this code.

GeorgeJahad commented on src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:
```diff
@@ -1005,6 +1031,140 @@ void handleMonitorPolicyEvent(MonitorPolicyEvent event) {
+
+    // In future, for policy monitors we should really get the effective policies based on the
+    // scope the monitor is configured for.
+    // Currently this step assumes all policy monitors are using globally scoped metadata.
```
I don't understand this comment.  Is it saying that non-global scope is not being used?

jjbuchan replied:
It's irrelevant now since I'm removing the logic around policy monitors.

This is the problem I mentioned where the policy monitor would be applied to many accounts (which should maybe have different metadata based on their account type) but since we were applying the metadata values at the monitor object level instead of when it is bound to the resource, all tenants would end up using the same "global" metadata for those monitors and not their "accountType" values if those were in place.

i.e. the monitor object would be created/stored with the global metadata values built into it.

GeorgeJahad commented on src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:
```diff
@@ -1005,6 +1031,140 @@ void handleMonitorPolicyEvent(MonitorPolicyEvent event) {
+        .collect(Collectors.toList());
+
+    // Remove existing bound monitors
+    // Policy monitors do not get rebound in this method, so can be skipped.
```
Is this because policy monitors don't use metadata?  If not, then why are they being updated below with the regular monitors?  what am i missing?

jjbuchan replied:
This is something else that is being removed based on our conversation from earlier.

In this version though, we do not unbind/rebind policy monitors all at once.  Instead we do similar to what is done when a policy monitor is updated - we send an event to policy mgmt to determine which tenants are affected by this change, then an event per tenant gets sent for monitor mgmt to consume and handle the rebinding per tenant rather than all at once.

...but as I say this is irrelevant now and policy monitors will not hit this method anymore.

jjbuchan commented:
The two new commits should address all the comments and what we talked about.

When removing the metadata from policy monitors it also exposed that the `interval` was not being set in policy monitor updates, so I've added that and tweaked the test to verify that works.  Other than that, the only other addition is that I've added a couple metrics to track errors.

Later on we'll need to add this file to all repos (or find a way to share it) - https://github.com/racker/salus-telemetry-ambassador/blob/master/src/main/java/com/rackspace/salus/telemetry/ambassador/config/MetricsConfig.java

itzg commented on src/main/java/com/rackspace/salus/monitor_management/services/MonitorManagement.java:
```diff
@@ -156,9 +159,14 @@ public MonitorManagement(
     this.jdbcTemplate = jdbcTemplate;
     this.labelMatchQuery = SpringResourceUtils.readContent("sql-queries/monitor_label_matching_query.sql");
+
+    monitorMetadataContentUpdateErrors = meterRegistry.counter("errors",
```
Thanks. I/we probably should have been adding a few of these each time we touched code.

jjbuchan replied:
Figured the easy way to start was with the exceptions that get thrown.
